pkg/objstore/azure/azure.go

```diff
@@ -0,0 +1,318 @@
+			return nil, err
+		}
+	} else {
+		level.Info(logger).Log("msg", "Container successfully created", "address", container)
+	}
+
+	bkt := &Bucket{
+		logger:       logger,
+		containerURL: container,
+		config:       &conf,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_objstore_azure_storage_operations_total",
+			Help:        "Total number of operations that were executed against an Azure storage account.",
+			ConstLabels: prometheus.Labels{"storage_account": conf.StorageAccountName},
+		}, []string{"operation"}),
+	}
+
+	if reg != nil {
+		reg.MustRegister(bkt.opsTotal)
+	}
+	return bkt, nil
+}
+
+// Iter calls f for each entry in the given directory. The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	level.Debug(b.logger).Log("msg", "Iterating through the dir", "dir", dir)
+	b.opsTotal.WithLabelValues(opObjectsList).Inc()
+	var prefix string
+	if dir == "" {
+		prefix = ""
+	} else if !strings.HasSuffix(dir, DirDelim) {
+		prefix = dir + DirDelim
+	} else {
+		prefix = dir
+	}
+
+	list, err := b.containerURL.ListBlobsHierarchySegment(ctx, blob.Marker{}, DirDelim, blob.ListBlobsSegmentOptions{
+		Prefix: prefix,
+	})
+	if err != nil {
+		level.Error(b.logger).Log("msg", "Cannot list blogs in directory %s", dir)
+		return err
+	}
+	var listNames []string
+
+	for _, blob := range list.Segment.BlobItems {
+		listNames = append(listNames, blob.Name)
+	}
+
+	for _, blobPrefix := range list.Segment.BlobPrefixes {
+		listNames = append(listNames, blobPrefix.Name)
+	}
+
+	for _, name := range listNames {
+		if err := f(name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	level.Debug(b.logger).Log("msg", "Check if err it's NotFoundError", "err", err)
+	if err == nil {
+		return false
+	}
+	switch errorCode := parseError(err.Error()); errorCode {
+	case "InvalidUri":
+		return true
+	case "BlobNotFound":
+		return true
+	default:
+		return false
+	}
+}
+
+func (b *Bucket) getBlobReader(ctx context.Context, name string, offset, length int64) (io.ReadCloser, error) {
+	level.Debug(b.logger).Log("msg", "Getting blob", "blob", name, "offset", offset, "length", length)
+	if len(name) == 0 {
+		return nil, errors.New("X-Ms-Error-Code: [EmptyContainerName]")
+	}
+	exists, err := b.Exists(ctx, name)
+
+	if !exists {
+		return nil, errors.New("X-Ms-Error-Code: [BlobNotFound]")
+	}
+
+	blobURL := getBlobURL(ctx, b.config.StorageAccountName, b.config.StorageAccountKey, b.config.ContainerName, name)
+
+	props, err := blobURL.GetProperties(ctx, blob.BlobAccessConditions{})
+	if err != nil {
+		level.Error(b.logger).Log("msg", "Cannot get properties for container", "address", name)
```

Comment:
Never log error and return in the same time. Can you double check against that in this whole PR? (: Basically this error will be logged anyway. Feel free to add `address` in `errors.Wrapf`